crates/topos-p2p/src/behaviour/grpc.rs

```diff
@@ -0,0 +1,480 @@
+
+        request_id
+    }
+
+    /// Try to open a connection with the given peer.
+    fn open_connection(&mut self, peer_id: &PeerId) -> OutboundConnection {
+        info!("Opening gRPC outbound connection to peer {peer_id}");
+
+        let (notifier, receiver) = oneshot::channel();
+        let request_id = self.next_request_id();
+
+        self.pending_outbound_connections
+            .entry(*peer_id)
+            .or_insert_with(|| OutboundConnectionRequest {
+                request_id,
+                notifier,
+            });
+
+        self.pending_events.push_back(ToSwarm::Dial {
+            opts: DialOpts::peer_id(*peer_id).build(),
+        });
+
+        OutboundConnection::Opening {
+            request_id,
+            receiver,
+        }
+    }
+
+    /// Handle the [`ConnectionEstablished`] event comming from the [`Swarm`]
+    /// and try to open a gRPC channel using a [`ConnectionHandler`].
+    fn on_connection_established(
+        &mut self,
+        ConnectionEstablished {
+            peer_id,
+            connection_id,
+            endpoint,
+            failed_addresses,
+            other_established,
+        }: ConnectionEstablished,
+    ) {
+        let address = match endpoint {
+            ConnectedPoint::Dialer { address, .. } => Some(address.clone()),
+            ConnectedPoint::Listener { .. } => None,
+        };
+
+        let connection = Connection {
+            id: connection_id,
+            address,
+            request_id: None,
+            channel: None,
+        };
+
+        self.connected.entry(peer_id).or_default().push(connection);
+
+        // TODO refactor this to better handle connection
+        // If there is no current established connection it means that it's the
+        // first connection with that peer
+        if other_established == 0 {
+            self.try_connect(&peer_id);
+        }
+    }
+
+    /// Starts the gRPC service if not already started
+    fn on_new_listener(&mut self, listener_id: ListenerId) {
+        if let Some(service) = self.service.take() {
+            let (tx, rx) = mpsc::unbounded_channel();
+            self.inbound_stream = Some(tx);
+            // TODO: Switch to serve_with_incoming_shutdown at some point
+            tokio::spawn(service.serve_with_incoming(proxy::GrpcProxy::new(rx)));
+            info!("New gRPC proxy start and listening on {listener_id:?}");
+        } else {
+            warn!("Tried to instantiate gRPC proxy but service is missing");
+        }
+    }
+
+    /// On [`ConnectionClosed`] we cleanup the `connected` state of the behaviour.
+    fn on_connection_closed(
+        &mut self,
+        ConnectionClosed {
+            peer_id,
+            connection_id,
+            endpoint,
+            handler,
+            remaining_established,
+        }: ConnectionClosed<<Self as NetworkBehaviour>::ConnectionHandler>,
+    ) {
+        warn!("Connection {connection_id} closed with peer {peer_id}");
```

Comment:
This is probably normal behaviour and a common occurrence. Better to use `debug`?